### ISSUE TITLE
Add a switch

### ISSUE THREAD

**Issue by samreid** [phetsims/circuit-construction-kit-common#88]
The design doc shows a switch that looks like this:

![image](https://cloud.githubusercontent.com/assets/679486/15102232/1a9669a2-155a-11e6-8916-9276e4c8b3ab.png)

With a description like so:
- Switch will be able to be changed by tapping or dragging to the alternate state (on/off)
- Switch can be rotated
- Switch connection wires should stretch “linearly” not at angles to switch

![image](https://cloud.githubusercontent.com/assets/679486/15102245/4e93aa44-155a-11e6-8e39-c4700d22c8f7.png)
- Switch has stretchable (linear) wires
- Switch rotatable
- drag or tap to activate switch
- Switch more plastic looking

**Comment by samreid** [comment]
I like in this design how the switch is stretchable, but I'm concerned that at its smallest size, it will be difficult to translate by dragging because these is very little area for dragging (the main area is taken up by the toggle button).

**Comment by samreid** [comment]
Tagging @arouinfar in case she wants to comment on these design issues.

**Comment by samreid** [comment]
Or perhaps the switch component can be used to close the switch as well as drag it?  That seems confusing (and perhaps difficult to implement), but may be useful in practice.

**Comment by samreid** [comment]
Also, it seems like having the switchable part look like it is resting on a metal plate may be confusing, if people think the plate is conducting.

**Comment by samreid** [comment]
Also, when we model the movement of electrons, we'll have to discuss how they look and behave in the switch segment.

**Comment by samreid** [comment]
Another UI component that may simplify things would be a "rotate" button that rotates the center segment of the switch.

**Comment by samreid** [comment]
Initial draft based on Wire committed in f62022b823d501636ee00becac6911830176ac40

**Comment by samreid** [comment]
I've added a checkbox that toggles the switch wire to be open/closed.  The ammeter and voltmeter will need custom code to handle this case.

**Comment by samreid** [comment]
I'm inclined to wait on further developing the switch until the design questions raised above are addressed, @arouinfar can you take a look?  Summarizing (and a few new points) here:
* Switch artwork not finalized
* at its smallest size, it will be difficult to translate by dragging because these is very little area for dragging (the main area is taken up by the toggle button).
* when you drag it out of the toolbox it drags instead of toggling the switch.
* Or perhaps the switch component can be used to close the switch as well as drag it? That seems confusing (and perhaps difficult to implement), but may be useful in practice.
* Or perhaps the switch component can be used to close the switch as well as drag it? That seems confusing (and perhaps difficult to implement), but may be useful in practice.
* it seems like having the switchable part look like it is resting on a metal plate may be confusing, if people think the plate is conducting.
* Also, when we model the movement of electrons, we'll have to discuss how they look and behave in the switch segment.  It could be complicated to show the electrons displacing with the open circuit.
* If we go with an On Off Switch or Check Box, should it always be visible, or only when the circuit is selected?

**Comment by arouinfar** [comment]
> I'm inclined to wait on further developing the switch

Yes, please hold off on this @samreid. As you've pointed out there is a lot left to iron out in the design of the switch, and you've raised quite a few points that merit further discussion.

I'll think about this a bit more, and we should discuss the design at the next design meeting.

**Comment by FelixINX** [comment]
Hello!

We are currently using the Circuit Con. Kit DC in our school, Java version. I just published [the first version online](http://phet-circuits.felixinx.me/circuit-construction-kit-dc_en.html) for our student.

We have to use the switch. As I can see in the code, a big part of the code is ready. Do you know when we will be able to use it? _Even if it is not the stable version._

Thanks!

**Comment by samreid** [comment]
@ariel-phet or @arouinfar can you please respond to @FelixINX?

**Comment by ariel-phet** [comment]
@FelixINX the time line for this element being ready has not crystallized yet. You had mentioned some interest in working on code in another issue. To evaluate that possibility I will need you to email me directly. Please provide me with an email address at which I can contact you.

**Comment by FelixINX** [comment]
Sure it is felix.desjardins2 [at] ggl.csmb.qc.ca

**Comment by ariel-phet** [comment]
@FelixINX I have sent you an email, and look forward to your reply

**Comment by arouinfar** [comment]
From 3/9/17 design meeting:

Switch could look something like its circuit symbol:
![image](https://cloud.githubusercontent.com/assets/8419308/23770838/eac5d3bc-04d1-11e7-8143-107e72160013.png)

From a development standpoint, keeping lifelike switch comparable to its schematic form is preferable.

Switch could be in a self-contained box that has a power button  inside. The power icon cues "on/off" and would invite users to tap the box.

**Comment by arouinfar** [comment]
If switch is tapped, its state is switched and it will be "selected". Text at the bottom of the screen could say "Switch is open/closed". Trash button will appear as usual.

**Comment by arouinfar** [comment]
Preliminary switch:
<img src="https://cloud.githubusercontent.com/assets/8419308/25026914/70f4e4c8-2066-11e7-8030-fdf97bc5097b.png" width="300">

In reality, the hard edges of the lock and key would grind against one another, so I tried rounding out the edges a bit. 
<img src="https://cloud.githubusercontent.com/assets/8419308/25053802/c0ba7e40-2116-11e7-9986-a2818f38ab38.png" width="300">

@samreid is there anything else you'll need for the switch?

**Comment by samreid** [comment]
@arouinfar I think we have everything we need.

**Comment by samreid** [comment]
Will there be animation or continuous motion for the switch?  Or just immediately jumping open and closed? If there is continuous motion, will there be continuous dragging, or only "tap to open/tap to close"?

**Comment by arouinfar** [comment]
@samreid tapping or clicking the switch should flip its state - no animation necessary. 

We had also talked about allowing users to continuously drag the switch arm, similar to Capacitor Lab: Basics. I think continuous dragging would be nice, but not critical. I'm also not sure how feasible it would be given that the CCK switch would be quite a bit smaller than the CL:B switch.

**Comment by samreid** [comment]
I'm skeptical about the continuous dragging working on iPads, so let's plan for now on the tap to toggle idea.

**Comment by arouinfar** [comment]
Sounds good @samreid.

**Comment by samreid** [comment]
> Text at the bottom of the screen could say "Switch is open/closed". Trash button will appear as usual.

Should there be a control at the bottom of the screen for opening and closing the switch?

**Comment by samreid** [comment]
@arouinfar I got most of the switch features working, including:

* lifelike render
* schematic render
* tap to toggle
* circuit physics
* resistance display when "show values" is selected, with infinity symbol
* icon in toolbox
* readout (circuit is open/closed) at the bottom of the screen

Known issues (will be moved to separate issues):
* the show values box moves up and down based on whether the switch is open or closed (unintentional and kind of buggy)
* the curvy curvy connectors
* there is no delete button at the bottom of the screen for selected switch

@arouinfar can you please review everything and comment?

http://www.colorado.edu/physics/phet/dev/html/circuit-construction-kit-dc/1.0.0-dev.20/circuit-construction-kit-dc_en.html

**Comment by arouinfar** [comment]
@samreid the switch looks great!! Tapping to open/close seems intuitive. 

Originally, the switch was sitting in a rounded rectangle, with the motivation being that it would help the switch feel like a contained component. However, after seeing the switch in action, I don't think it's necessary. 

I find that the flat appearance of the switch looks a bit odd compared to the other components. @samreid can you apply a slight gradient to the switch? 

I'm also not sure if the switch should be layered above or below the solder. I think it would look better if layered below, but I know the current logic is that fixed-width components are layered above the solder.

**Comment by samreid** [comment]
Sounds good, I'll move your two points to new issues.
